diff --git a/MODULE.bazel b/MODULE.bazel
index 25456185..4c063db5 100644
--- a/MODULE.bazel
+++ b/MODULE.bazel
@@ -102,6 +102,12 @@ go_deps.from_file(go_mod = "//:go.mod")
 use_repo(go_deps, "com_github_inspirer_textmapper")
 
 # Python dependencies.
+python = use_extension("@rules_python//python/extensions:python.bzl", "python")
+python.toolchain(
+    is_default = True,
+    python_version = "3.11",
+)
+
 pip = use_extension("@rules_python//python/extensions:pip.bzl", "pip")
 pip.parse(
     hub_name = "pypi",
diff --git a/googlesql/analyzer/BUILD b/googlesql/analyzer/BUILD
index e057a93d..b26784a8 100644
--- a/googlesql/analyzer/BUILD
+++ b/googlesql/analyzer/BUILD
@@ -21,7 +21,7 @@ load("@rules_cc//cc:cc_test.bzl", "cc_test")
 load(":builddefs.bzl", "gen_analyzer_test")
 
 package(
-    default_visibility = ["//googlesql/base:googlesql_implementation"],
+    default_visibility = ["//visibility:public"],
 )
 
 filegroup(
diff --git a/googlesql/analyzer/graph_query_resolver.cc b/googlesql/analyzer/graph_query_resolver.cc
index c8f3edbd..97ccdf33 100644
--- a/googlesql/analyzer/graph_query_resolver.cc
+++ b/googlesql/analyzer/graph_query_resolver.cc
@@ -3693,7 +3693,7 @@ GraphTableQueryResolver::ResolveGqlOperatorList(
     scan_list.pop_back();
     GOOGLESQL_ASSIGN_OR_RETURN(std::unique_ptr<const ResolvedFinishScan> finish_scan,
                      ResolvedFinishScanBuilder()
-                         .set_column_list({})
+                         .set_column_list(std::vector<ResolvedColumn>{})
                          .set_input_scan(std::move(last_scan))
                          .Build());
     scan_list.push_back(std::move(finish_scan));
diff --git a/googlesql/analyzer/name_scope.cc b/googlesql/analyzer/name_scope.cc
index 0931909e..38fd61f8 100644
--- a/googlesql/analyzer/name_scope.cc
+++ b/googlesql/analyzer/name_scope.cc
@@ -1882,7 +1882,7 @@ NameList::AddRangeVariableInWrappingNameList(
   // variables, including for value tables, so we use `flatten_to_table`
   // which drops range variables.
   GOOGLESQL_RETURN_IF_ERROR(range_variable_name_list->MergeFrom(
-      *original_name_list, ast_location, {.flatten_to_table = true}));
+      *original_name_list, ast_location, MergeOptions{.flatten_to_table = true}));
 
   auto wrapper_name_list = std::make_shared<NameList>();
   GOOGLESQL_RETURN_IF_ERROR(
diff --git a/googlesql/analyzer/resolver_expr.cc b/googlesql/analyzer/resolver_expr.cc
index cd73c57b..6ebed2c4 100644
--- a/googlesql/analyzer/resolver_expr.cc
+++ b/googlesql/analyzer/resolver_expr.cc
@@ -3131,7 +3131,7 @@ class SystemVariableConstant final : public Constant {
                          const Type* type)
       : Constant(name_path), type_(type) {}
 
-  const Type* type() const override { return type_; }
+  const googlesql::Type* type() const override { return type_; }
   std::string DebugString() const override { return FullName(); }
   std::string ConstantValueDebugString() const override { return "<N/A>"; }
 
diff --git a/googlesql/analyzer/substitute.cc b/googlesql/analyzer/substitute.cc
index 624a0b58..169a7307 100644
--- a/googlesql/analyzer/substitute.cc
+++ b/googlesql/analyzer/substitute.cc
@@ -227,15 +227,12 @@ class VariableReplacementInserter : public ResolvedASTDeepCopyVisitor {
     GOOGLESQL_RET_CHECK_FAIL() << "Unexpected node type: "
                      << ResolvedNodeKind_Name(node->node_kind());
   }
-  template <>
   absl::Status CopyVisitNode(const ResolvedSubqueryExpr* node) {
     return CopyVisitResolvedSubqueryExpr(node);
   }
-  template <>
   absl::Status CopyVisitNode(const ResolvedInlineLambda* node) {
     return CopyVisitResolvedInlineLambda(node);
   }
-  template <>
   absl::Status CopyVisitNode(const ResolvedJoinScan* node) {
     return CopyVisitResolvedJoinScan(node);
   }
diff --git a/googlesql/base/BUILD b/googlesql/base/BUILD
index da713437..14b55907 100644
--- a/googlesql/base/BUILD
+++ b/googlesql/base/BUILD
@@ -20,7 +20,7 @@ load("@rules_cc//cc:cc_test.bzl", "cc_test")
 
 licenses(["notice"])
 
-package(default_visibility = [":googlesql_implementation"])
+package(default_visibility = ["//visibility:public"])
 
 package_group(
     name = "googlesql_implementation",
@@ -82,6 +82,15 @@ cc_library(
     deps = [":check"],
 )
 
+cc_library(
+    name = "no_destructor",
+    hdrs = ["no_destructor.h"],
+    visibility = ["//visibility:public"],
+    deps = [
+        "@com_google_absl//absl/base:no_destructor",
+    ],
+)
+
 cc_library(
     name = "arena",
     srcs = ["arena.cc"],
diff --git a/googlesql/base/logging.h b/googlesql/base/logging.h
index 730ccdcb..46fe06b0 100644
--- a/googlesql/base/logging.h
+++ b/googlesql/base/logging.h
@@ -59,6 +59,17 @@ inline void GoogleSqlMakeCheckOpValueString(std::ostream *os, const T &v) {
   (*os) << v;
 }
 
+// This overloading is implemented to address the compilation issue when trying to log unique_ptr types
+// At the moment, we are not providing any specific implementation for handling unique_ptr types.
+template <typename T>
+inline void GoogleSqlMakeCheckOpValueString(std::ostream* os, const std::unique_ptr<T>& v) {
+  if (v == nullptr) {
+    (*os) << "nullptr";
+  } else {
+    (*os) << v.get();
+  }
+}
+
 // Overrides for char types provide readable values for unprintable
 // characters.
 template <>
diff --git a/googlesql/base/no_destructor.h b/googlesql/base/no_destructor.h
new file mode 100644
index 00000000..2d476a0e
--- /dev/null
+++ b/googlesql/base/no_destructor.h
@@ -0,0 +1,28 @@
+// Copyright 2026 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifndef GOOGLESQL_BASE_NO_DESTRUCTOR_H_
+#define GOOGLESQL_BASE_NO_DESTRUCTOR_H_
+
+#include "absl/base/no_destructor.h"
+
+namespace googlesql_base {
+
+template <typename T>
+using NoDestructor = absl::NoDestructor<T>;
+
+}  // namespace googlesql_base
+
+#endif  // GOOGLESQL_BASE_NO_DESTRUCTOR_H_
diff --git a/googlesql/base/testing/BUILD b/googlesql/base/testing/BUILD
index a5be680b..7c0b63ed 100644
--- a/googlesql/base/testing/BUILD
+++ b/googlesql/base/testing/BUILD
@@ -19,7 +19,7 @@ load("@rules_cc//cc:cc_test.bzl", "cc_test")
 
 licenses(["notice"])
 
-package(default_visibility = ["//googlesql/base:googlesql_implementation"])
+package(default_visibility = ["//visibility:public"])
 
 # A drop in replacement for gtest_main that parsers absl flags
 cc_library(
@@ -60,6 +60,7 @@ cc_library(
         "@com_google_absl//absl/strings",
         "@com_google_absl//absl/strings:string_view",
         "@com_google_googletest//:gtest",
+        "@com_github_grpc_grpc//:grpc++",
     ],
 )
 
@@ -74,6 +75,7 @@ cc_test(
         "@com_google_absl//absl/status",
         "@com_google_absl//absl/status:statusor",
         "@com_google_googletest//:gtest",
+        "@com_github_grpc_grpc//:grpc++",
     ],
 )
 
diff --git a/googlesql/common/BUILD b/googlesql/common/BUILD
index 12e38abe..7e18a475 100644
--- a/googlesql/common/BUILD
+++ b/googlesql/common/BUILD
@@ -18,7 +18,7 @@ load("@rules_cc//cc:cc_test.bzl", "cc_test")
 load(":builddefs.bzl", "gen_module_test")
 
 package(
-    default_visibility = ["//googlesql/base:googlesql_implementation"],
+    default_visibility = ["//visibility:public"],
     features = ["parse_headers"],
 )
 
diff --git a/googlesql/common/internal_value.h b/googlesql/common/internal_value.h
index 25d5c27a..01ab943c 100644
--- a/googlesql/common/internal_value.h
+++ b/googlesql/common/internal_value.h
@@ -116,7 +116,7 @@ class InternalValue {
   static std::string FormatInternal(const Value& x,
                                     bool include_array_ordereness
   ) {
-    return x.FormatInternal({
+    return x.FormatInternal(Type::FormatValueContentOptions{
         .force_type_at_top_level = true,
         .include_array_ordereness = include_array_ordereness,
         .indent = 0,
diff --git a/googlesql/parser/BUILD b/googlesql/parser/BUILD
index b1ca7a50..863b2471 100644
--- a/googlesql/parser/BUILD
+++ b/googlesql/parser/BUILD
@@ -28,7 +28,7 @@ load("//bazel:textmapper.bzl", "tm_syntax")
 load(":builddefs.bzl", "gen_parser_test")
 
 package(
-    default_visibility = ["//googlesql/base:googlesql_implementation"],
+    default_visibility = ["//visibility:public"],
 )
 
 genrule(
diff --git a/googlesql/public/coercer.cc b/googlesql/public/coercer.cc
index 930ac280..5761c002 100644
--- a/googlesql/public/coercer.cc
+++ b/googlesql/public/coercer.cc
@@ -154,7 +154,7 @@ class TypeSuperTypes {
     return false;
   }
 
-  const Type* type() const { return type_; }
+  const googlesql::Type* type() const { return type_; }
   TypeListView supertypes() const { return supertypes_; }
 
   std::vector<const Type*> ToVector() const {
diff --git a/googlesql/public/constant.h b/googlesql/public/constant.h
index 946183b0..177a5dfd 100644
--- a/googlesql/public/constant.h
+++ b/googlesql/public/constant.h
@@ -80,7 +80,7 @@ class Constant {
   const std::vector<std::string>& name_path() const { return name_path_; }
 
   // Returns the type of this Constant.
-  virtual const Type* type() const = 0;
+  virtual const googlesql::Type* type() const = 0;
 
   // Returns whether or not this Constant is a specific constant interface or
   // implementation.
diff --git a/googlesql/public/function_signature.h b/googlesql/public/function_signature.h
index 2655a3cc..3aac8e2c 100644
--- a/googlesql/public/function_signature.h
+++ b/googlesql/public/function_signature.h
@@ -765,7 +765,7 @@ class FunctionArgumentType {
   // Returns NULL if kind_ is not ARG_KIND_EXPR_FIXED or ARG_KIND_LAMBDA.
   // If kind_ is ARG_KIND_LAMBDA, returns the type of lambda body type, which
   // could be NULL if the body type is templated.
-  const Type* type() const { return type_; }
+  const googlesql::Type* type() const { return type_; }
 
   SignatureArgumentKind kind() const { return kind_; }
 
diff --git a/googlesql/public/input_argument_type.h b/googlesql/public/input_argument_type.h
index 5dc60a10..ff79ef86 100644
--- a/googlesql/public/input_argument_type.h
+++ b/googlesql/public/input_argument_type.h
@@ -138,7 +138,7 @@ class InputArgumentType {
   ~InputArgumentType() = default;
 
   // This may return nullptr (such as for lambda).
-  const Type* type() const { return annotated_type().type; }
+  const googlesql::Type* type() const { return annotated_type().type; }
   const AnnotationMap* annotation_map() const {
     return annotated_type().annotation_map;
   }
diff --git a/googlesql/public/property_graph.h b/googlesql/public/property_graph.h
index 6f23b6de..5a1c730b 100644
--- a/googlesql/public/property_graph.h
+++ b/googlesql/public/property_graph.h
@@ -378,7 +378,7 @@ class GraphPropertyDeclaration {
     return ::googlesql::FullName(PropertyGraphNamePath(), Name());
   }
 
-  virtual const Type* Type() const = 0;
+  virtual const googlesql::Type* Type() const = 0;
 
   // TODO: This should be made a pure virtual function once all subclasses
   // implement it.
diff --git a/googlesql/public/simple_catalog.cc b/googlesql/public/simple_catalog.cc
index f909948a..5dd56444 100644
--- a/googlesql/public/simple_catalog.cc
+++ b/googlesql/public/simple_catalog.cc
@@ -1767,7 +1767,7 @@ SimpleTable::SimpleTable(absl::string_view name,
 // to something named 'value' is confusing there.
 SimpleTable::SimpleTable(absl::string_view name, const Type* row_type,
                          const int64_t id)
-    : SimpleTable(name, {{"value", row_type}}, id) {
+    : SimpleTable(name, {NameAndType({"value", row_type})}, id) {
   is_value_table_ = true;
 }
 
diff --git a/googlesql/public/simple_catalog.h b/googlesql/public/simple_catalog.h
index c93ae868..a1dc30b5 100644
--- a/googlesql/public/simple_catalog.h
+++ b/googlesql/public/simple_catalog.h
@@ -1257,7 +1257,7 @@ class SimpleConstant : public Constant {
       const SimpleConstantProto& simple_constant_proto,
       const TypeDeserializer& type_deserializer);
 
-  const Type* type() const override { return value_.type(); }
+  const googlesql::Type* type() const override { return value_.type(); }
 
   const Value& value() const { return value_; }
 
diff --git a/googlesql/public/sql_constant.h b/googlesql/public/sql_constant.h
index fa88344f..69defd3b 100644
--- a/googlesql/public/sql_constant.h
+++ b/googlesql/public/sql_constant.h
@@ -60,7 +60,7 @@ class SQLConstant : public Constant {
 
   // Returns the Type of the resolved Constant based on its resolved
   // expression type.
-  const Type* type() const override {
+  const googlesql::Type* type() const override {
     return constant_expression()->type();
   }
 
diff --git a/googlesql/public/types/BUILD b/googlesql/public/types/BUILD
index 512ab4f6..4be021d8 100644
--- a/googlesql/public/types/BUILD
+++ b/googlesql/public/types/BUILD
@@ -17,7 +17,7 @@
 load("@rules_cc//cc:cc_library.bzl", "cc_library")
 load("@rules_cc//cc:cc_test.bzl", "cc_test")
 
-package(default_visibility = ["//googlesql/base:googlesql_implementation"])
+package(default_visibility = ["//visibility:public"])
 
 filegroup(
     name = "type_headers",
diff --git a/googlesql/public/types/array_type.cc b/googlesql/public/types/array_type.cc
index a2b55415..e7a54798 100644
--- a/googlesql/public/types/array_type.cc
+++ b/googlesql/public/types/array_type.cc
@@ -121,7 +121,7 @@ struct HashableValueContentContainerElementIgnoringFloat {
         break;
       }
       case TYPE_PROTO: {
-        bool has_floating_point_fields = v.type->HasFloatingPointFields();
+        bool has_floating_point_fields = v.type->has_floating_point_fields();
         if (has_floating_point_fields) {
           h = H::combine(std::move(h), TYPE_PROTO, has_floating_point_fields);
           break;
diff --git a/googlesql/public/types/type.h b/googlesql/public/types/type.h
index cbf8807d..2721a679 100644
--- a/googlesql/public/types/type.h
+++ b/googlesql/public/types/type.h
@@ -701,6 +701,8 @@ class Type {
     std::optional<int64_t> file_descriptor_sets_max_size_bytes = std::nullopt;
   };
 
+  bool has_floating_point_fields() const { return HasFloatingPointFields(); }
+
  protected:
   Type(const TypeFactoryBase& factory, TypeKind kind);
   virtual ~Type() = default;
diff --git a/googlesql/public/value.cc b/googlesql/public/value.cc
index d5b35c63..dc236d12 100644
--- a/googlesql/public/value.cc
+++ b/googlesql/public/value.cc
@@ -1243,7 +1243,7 @@ std::string Value::DebugString(bool verbose) const {
 
 // Format will wrap arrays and structs.
 std::string Value::Format(bool print_top_level_type) const {
-  return FormatInternal(
+  return FormatInternal(Type::FormatValueContentOptions
       {.force_type_at_top_level = print_top_level_type, .indent = 0});
 }
 
@@ -1511,7 +1511,7 @@ std::string Value::FormatInternal(
     std::vector<std::string> element_strings(elements().size());
     for (int i = 0; i < elements().size(); ++i) {
       element_strings[i] =
-          elements()[i].FormatInternal(options.IncreaseIndent());
+          elements()[i].FormatInternal(Type::FormatValueContentOptions{options.IncreaseIndent()});
     }
     // Sanitize any '$' characters before creating substitution template. "$$"
     // is replaced by "$" in the output from absl::Substitute.
@@ -1553,7 +1553,7 @@ std::string Value::FormatInternal(
     const StructType* struct_type = type()->AsStruct();
     std::vector<std::string> field_strings(struct_type->num_fields());
     for (int i = 0; i < struct_type->num_fields(); i++) {
-      field_strings[i] = fields()[i].FormatInternal(options.IncreaseIndent());
+      field_strings[i] = fields()[i].FormatInternal(Type::FormatValueContentOptions{options.IncreaseIndent()});
     }
     // Sanitize any '$' characters before creating substitution template. "$$"
     // is replaced by "$" in the output from absl::Substitute.
@@ -1599,9 +1599,9 @@ std::string Value::FormatInternal(
     }
     std::vector<std::string> boundaries_strings;
     boundaries_strings.push_back(
-        start().FormatInternal(options.IncreaseIndent()));
+        start().FormatInternal(Type::FormatValueContentOptions{options.IncreaseIndent()}));
     boundaries_strings.push_back(
-        end().FormatInternal(options.IncreaseIndent()));
+        end().FormatInternal(Type::FormatValueContentOptions{options.IncreaseIndent()}));
     // Sanitize any '$' characters before creating substitution template. "$$"
     // is replaced by "$" in the output from absl::Substitute.
     std::string templ =
diff --git a/googlesql/public/value.h b/googlesql/public/value.h
index 9c0006fc..94eb14d1 100644
--- a/googlesql/public/value.h
+++ b/googlesql/public/value.h
@@ -124,7 +124,7 @@ class Value {
 
   // Returns the type of the value.
   // Requires (crashes otherwise): is_valid()
-  const Type* type() const;
+  const googlesql::Type* type() const;
 
   // Returns the type kind of the value. Same as type()->type_kind() but in some
   // cases can be a bit more efficient.
@@ -1348,7 +1348,7 @@ class Value {
 
     // Returns a pointer to Value's Type. Requires is_valid(). If TypeKind is
     // stored in the Metadata, Type pointer is obtained from static TypeFactory.
-    const Type* type() const;
+    const googlesql::Type* type() const;
 
     // Returns true, if instance stores pointer to a Type and false if type's
     // kind.
diff --git a/googlesql/public/value_inl.h b/googlesql/public/value_inl.h
index f9aedb86..6c7e2463 100644
--- a/googlesql/public/value_inl.h
+++ b/googlesql/public/value_inl.h
@@ -1310,14 +1310,14 @@ class Value::Metadata::Content {
   // TODO: wait for fixed clang-format
   // clang-format on
 
-  const Type* type() const {
+  const googlesql::Type* type() const {
     return reinterpret_cast<const Type*>(type_ & kTypeMask);
   }
   int32_t value_extended_content() const { return value_extended_content_; }
   bool is_null() const { return type_ & kIsNullTag; }
   bool preserves_order() const { return type_ & kPreserverOrderTag; }
   bool has_type_pointer() const { return type_ & kHasTypeTag; }
-  uint64_t raw_type() const { return type_; }
+  constexpr uint64_t raw_type() const { return type_; }
 
   friend constexpr Value::Metadata::Metadata(TypeKind kind, bool is_null,
                                              bool preserves_order,
diff --git a/googlesql/reference_impl/operator.h b/googlesql/reference_impl/operator.h
index a432666c..6b17e767 100644
--- a/googlesql/reference_impl/operator.h
+++ b/googlesql/reference_impl/operator.h
@@ -242,7 +242,7 @@ class ExprArg : public AlgebraArg {
 
   ~ExprArg() override = default;
 
-  const Type* type() const { return type_; }
+  const googlesql::Type* type() const { return type_; }
 
  private:
   const Type* type_;
diff --git a/googlesql/resolved_ast/BUILD b/googlesql/resolved_ast/BUILD
index e036906c..f93565ba 100644
--- a/googlesql/resolved_ast/BUILD
+++ b/googlesql/resolved_ast/BUILD
@@ -210,6 +210,7 @@ cc_library(
         "-O3",
         "-Wframe-larger-than=4000",
     ],
+    visibility = ["//visibility:public"],
     deps = [
         ":resolved_ast",
         ":resolved_ast_builder",
diff --git a/googlesql/resolved_ast/resolved_column.h b/googlesql/resolved_ast/resolved_column.h
index d5e484b8..dc9c7eca 100644
--- a/googlesql/resolved_ast/resolved_column.h
+++ b/googlesql/resolved_ast/resolved_column.h
@@ -121,7 +121,7 @@ class ResolvedColumn {
   IdString table_name_id() const { return table_name_; }
   IdString name_id() const { return name_; }
 
-  const Type* type() const { return annotated_type_.type; }
+  const googlesql::Type* type() const { return annotated_type_.type; }
 
   const AnnotationMap* /*absl_nullable*/ type_annotation_map() const {
     return annotated_type_.annotation_map;
diff --git a/googlesql/resolved_ast/rewrite_utils.cc b/googlesql/resolved_ast/rewrite_utils.cc
index aee35118..9ec9e236 100644
--- a/googlesql/resolved_ast/rewrite_utils.cc
+++ b/googlesql/resolved_ast/rewrite_utils.cc
@@ -70,6 +70,7 @@ namespace googlesql {
 namespace {
 
 struct ConcreteArgument {
+  ConcreteArgument(const Type* t, size_t num) : type(t), num_occurrences(num) {}
   const Type* type;
   size_t num_occurrences;
 };
diff --git a/googlesql/testing/test_value.h b/googlesql/testing/test_value.h
index 6d59d274..cc8d1398 100644
--- a/googlesql/testing/test_value.h
+++ b/googlesql/testing/test_value.h
@@ -110,7 +110,7 @@ class ValueConstructor {
       : v_(v) {}
 
   const Value& get() const { return v_; }
-  const Type* type() const { return v_.type(); }
+  const googlesql::Type* type() const { return v_.type(); }
 
   static std::vector<Value> ToValues(absl::Span<const ValueConstructor> slice) {
     std::vector<Value> values;
