diff --git a/bazel/grpc_extra_deps.bzl b/bazel/grpc_extra_deps.bzl
index 4a8e18ce17..cfbfd15678 100644
--- a/bazel/grpc_extra_deps.bzl
+++ b/bazel/grpc_extra_deps.bzl
@@ -64,9 +64,6 @@ def grpc_extra_deps(ignore_version_differences = False):
 
     api_dependencies()
 
-    go_rules_dependencies()
-    go_register_toolchains(version = "1.22.5")
-    gazelle_dependencies()
 
     # Pull-in the go 3rd party dependencies for protoc_gen_validate, which is
     # needed for building C++ xDS protos
diff --git a/include/grpcpp/impl/status.h b/include/grpcpp/impl/status.h
index 94c0e80a9b..23253ffee1 100644
--- a/include/grpcpp/impl/status.h
+++ b/include/grpcpp/impl/status.h
@@ -21,6 +21,7 @@
 
 // IWYU pragma: private, include <grpcpp/support/status.h>
 
+#include "absl/status/status.h"
 #include <grpc/status.h>
 #include <grpc/support/port_platform.h>
 #include <grpcpp/support/config.h>
@@ -98,6 +99,10 @@ class GRPC_MUST_USE_RESULT_WHEN_USE_STRICT_WARNING GRPCXX_DLL Status {
   Status(StatusCode code, const std::string& error_message)
       : code_(code), error_message_(error_message) {}
 
+  operator absl::Status() const& {
+    return absl::Status(static_cast<absl::StatusCode>(code_), error_message_);
+  }
+
   /// Construct an instance with \a code,  \a error_message and
   /// \a error_details. It is an error to construct an OK status with non-empty
   /// \a error_message and/or \a error_details.
diff --git a/src/core/util/glob.cc b/src/core/util/glob.cc
index 1b1c16e23d..22e54259f0 100644
--- a/src/core/util/glob.cc
+++ b/src/core/util/glob.cc
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <algorithm>
+
 #include "absl/strings/string_view.h"
 
 namespace grpc_core {
