--- zetasql/base/BUILD
+++ zetasql/base/BUILD
@@ -20,1 +20,1 @@
-    default_visibility = ["//:__subpackages__"],
+    default_visibility = ["//visibility:public"],

--- zetasql/base/testing/BUILD
+++ zetasql/base/testing/BUILD
@@ -19,1 +19,1 @@
-package(default_visibility = ["//:__subpackages__"])
+package(default_visibility = ["//visibility:public"])

--- zetasql/base/status.h
+++ zetasql/base/status.h
@@ -27,0 +28 @@
+#include "grpcpp/support/status.h"
@@ -89,0 +91,4 @@
+
+  Status(const grpc::Status& status)
+      : Status(static_cast<StatusCode>(status.error_code()),
+               status.error_message()) {}

--- zetasql/base/BUILD
+++ zetasql/base/BUILD
@@ -168,0 +168,1 @@
+        "@com_github_grpc_grpc//:grpc++",

--- bazel/icu4c-64_2.patch
+++ bazel/icu4c-64_2.patch
@@ -12,0 +12,6 @@
+
+--- source/config/mh-darwin
++++ source/config/mh-darwin
+@@ -22 +22 @@
+-ARFLAGS += -c
++ARFLAGS += -crs

--- zetasql/base/arena.cc
+++ zetasql/base/arena.cc
@@ -29 +29 @@
-#include <malloc.h>
+#include <stdlib.h>

--- bazel/m4.patch
+++ bazel/m4.patch
@@ -115,0 +115,6 @@
+
+--- lib/vasnprintf.c
++++ lib/vasnprintf.c
+@@ -4861 +4861 @@
+-# if !(((__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 3)) && !defined __UCLIBC__) || ((defined _WIN32 || defined __WIN32__) && ! defined __CYGWIN__))
++# if !defined(__APPLE__) && !(((__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 3)) && !defined __UCLIBC__) || ((defined _WIN32 || defined __WIN32__) && ! defined __CYGWIN__))

--- bazel/flex.patch
+++ bazel/flex.patch
@@ -3 +3 @@
-@@ -47,6 +47,7 @@
+@@ -47,6 +47,9 @@
@@ -6,0 +7 @@
++#if !defined(__APPLE__)
@@ -7,0 +9 @@
++#endif
