--- zetasql/base/BUILD
+++ zetasql/base/BUILD
@@ -20,1 +20,1 @@
-    default_visibility = [":zetasql_implementation"],
+    default_visibility = ["//visibility:public"],

--- zetasql/base/testing/BUILD
+++ zetasql/base/testing/BUILD
@@ -19,1 +19,1 @@
-package(default_visibility = ["//zetasql/base:zetasql_implementation"])
+package(default_visibility = ["//visibility:public"])

--- bazel/zetasql_deps_step_2.bzl
+++ bazel/zetasql_deps_step_2.bzl
@@ -460,13 +460,12 @@ exports_files(["data"])
             http_archive(
                 name = "m4",
                 build_file_content = all_content,
-                strip_prefix = "m4-1.4.18",
-                sha256 = "ab2633921a5cd38e48797bf5521ad259bdc4b979078034a3b790d7fec5493fab",
+                strip_prefix = "m4-1.4.19",
+                sha256 = "3be4a26d825ffdfda52a56fc43246456989a3630093cced3fbddf4771ee58a70",
                 urls = [
-                    "https://ftp.gnu.org/gnu/m4/m4-1.4.18.tar.gz",
-                    "https://mirrors.kernel.org/gnu/m4/m4-1.4.18.tar.gz",
+                    "https://ftp.gnu.org/gnu/m4/m4-1.4.19.tar.gz",
+                    "https://mirrors.kernel.org/gnu/m4/m4-1.4.19.tar.gz",
                 ],
-                patches = ["@com_google_zetasql//bazel:m4.patch"],
             )
 
         http_archive(