--- zetasql/base/BUILD
+++ zetasql/base/BUILD
@@ -20,1 +20,1 @@
-    default_visibility = [":zetasql_implementation"],
+    default_visibility = ["//visibility:public"],

--- zetasql/base/testing/BUILD
+++ zetasql/base/testing/BUILD
@@ -19,1 +19,1 @@
-package(default_visibility = ["//zetasql/base:zetasql_implementation"])
+package(default_visibility = ["//visibility:public"])

--- zetasql/parser/gen_parse_tree.py
+++ zetasql/parser/gen_parse_tree.py
@@ -23,7 +23,7 @@ Still a work in progress.
 
 """
 
-import collections
+import collections.abc
 import enum
 import operator
 import re
 
--- zetasql/resolved_ast/BUILD
+++ zetasql/resolved_ast/BUILD
@@ -33,6 +33,7 @@ py_binary(
         "@markupsafe//:markupsafe",
         "//zetasql/parser:generator_utils",
     ],
+    python_version = "PY3",
 )
 
 gen_resolved_ast_files(
